### PR TITLE
[HybridEP] Add dense top-k routing scan path

### PR DESCRIPTION
**Pull request by harryzhou2000**
## Summary
- Add dense int16 top-k routing input for HybridEP metadata preprocessing.
- Optimize dense scan with local-expert and rank bitsets.
- Improve HybridEP test diagnostics and benchmark labels, including with/without-probs paths.

## Performance
Primary impact is scan preprocessing: dense scan is significantly faster on the target dense top-k path. End-to-end dispatch/combine bandwidth is effectively unchanged versus upstream hybrid-ep in single-node B300 tests, with no observed regression.

## Testing
- Built on B300 with `TORCH_CUDA_ARCH_LIST="10.0;10.3+PTX"`.
- `tests/test_hybrid_ep.py --num-processes 8` on B300 for default config.
- Same test for H=512/E=8/K=8 and H=4096/E=8/K=8.
- BF16 and FP8 correctness passed, including dense top-k scan and dense top-k scan+permute paths.
- Re-ran updated test script against upstream hybrid-ep; dense-routing checks skip cleanly when unsupported.

### CODE REVIEW

File: tests/test_hybrid_ep.py [hunk 1]
```diff
@@ -246,6 +344,76 @@ def test_hybrid_ep_correctness(buffer: deep_ep.HybridEPBuffer, ref: TorchRef, us
             )
             print(f'  {api_name}: PASS', flush=True)
 
+    # Dense top-k routing with permute correctness check. This exercises scan
+    # with dense topk_idx input and enable_permute=True, i.e. the path that
+    # produces dense_chunk_layout and dense_to_expert_map.
+    if supports_dense_permute:
```
Comment (Autumn1998) [inline tests/test_hybrid_ep.py]: This part seems that nothing different with above, maybe we could use `for supports_dense_permute in [False, True]:` ?

File: deep_ep/hybrid_ep_buffer.py [hunk 2]
```diff
@@ -177,6 +195,7 @@ def dispatch(
         num_dispatched_tokens_tensor: torch.Tensor = None,
         num_dispatched_tokens: int = None,
         handle: tuple = None,
+        dense_routing: bool = False,
```
Comment (Autumn1998) [inline deep_ep/hybrid_ep_buffer.py]: Can we infer dense routing from the input? That way, we can avoid modifying the public-facing API.

File: tests/test_hybrid_ep.py (hunk 1 above)
Comment (harryzhou2000) [reply tests/test_hybrid_ep.py]: Addressed in the latest push. I refactored the permute correctness coverage into a shared routing-mode loop: sparse bool routing always runs, and dense topk_idx routing is added to the same loop when the installed HybridEP exposes dense top-k metadata. This keeps the sparse and dense checks using the same dispatch/ref/combine logic while still skipping dense checks cleanly on old hybrid-ep.

File: deep_ep/hybrid_ep_buffer.py (hunk 2 above)
Comment (harryzhou2000) [reply deep_ep/hybrid_ep_buffer.py]: Agreed and addressed. The public dense_routing flag has been removed. HybridEP now infers dense top-k routing when topk_idx is provided and routing_map is omitted; if routing_map is provided, it keeps the sparse bool path.

File: deep_ep/hybrid_ep_buffer.py [hunk 3]
```diff
@@ -186,33 +204,44 @@ def dispatch(
...
-            if topk_idx is not None:
-                routing_map, probs = indices_to_map(
+            topk = 0
+            routing_data = routing_map
+        elif topk_idx is not None:
+            assert num_of_experts is not None, "num_of_experts is required with topk_idx"
+            assert num_of_experts <= INT16_EXPERT_LIMIT, \
+                "dense topk_idx routing uses int16 expert IDs, so num_of_experts must be <= 32768"
+            topk = topk_idx.size(-1)
+            routing_data = topk_idx.to(torch.int16).contiguous()
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply deep_ep/hybrid_ep_buffer.py]: Addressed in 936e059. topk_idx-only calls now use dense int16 routing only when the dense limits are satisfied; otherwise they fall back to the pre-existing sparse indices_to_map path. I also validated this with a targeted NUM_LOCAL_EXPERTS=513 fallback check that asserted handle.config.topk == 0.

File: deep_ep/hybrid_ep_buffer.py [hunk 4]
```diff
@@ -186,33 +204,44 @@ def dispatch(
...
+            if probs is None and topk_weights is not None:
+                probs = dense_indices_to_probs(
                     topk_idx, topk_weights, num_of_tokens, num_of_experts
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply deep_ep/hybrid_ep_buffer.py]: Addressed in 936e059. _prepare_routing_data now validates num_of_experts against the configured HybridEP layout before constructing dense/sparse probabilities, so mismatched expert counts fail before GPU dispatch can index an incorrectly shaped probs tensor.

File: deep_ep/hybrid_ep_buffer.py [hunk 5]
```diff
@@ -186,33 +204,44 @@ def dispatch(
 
         Backward direction:
         combine_in_backward <- local_unpermute -> expert_mlp -> local_permute -> dispatch_in_backward
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply deep_ep/hybrid_ep_buffer.py:286]: Addressed in 936e059. topk_idx-only calls now preserve the sparse-map fallback outside dense limits. If routing_map and topk_idx are both supplied, routing_map still takes precedence, and that precedence is now documented in both dispatch docstrings.

File: deep_ep/hybrid_ep_buffer.py [hunk 6]
```diff
@@ -186,33 +204,44 @@ def dispatch(
...
+
+        When routing_map is omitted, topk_idx is passed directly as int16 (skipping indices_to_map).
+        This reduces allgather size from T*E_total to T*K*2 bytes.
+        Dropped tokens should use -1 as sentinel (naturally ignored by range checks in the kernel).
         """
         num_of_tokens, hidden_dim = hidden.shape
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply deep_ep/hybrid_ep_buffer.py:295]: Addressed in 936e059. The dense path now documents the expert-id contract at the Python boundary: valid IDs are [0, num_of_experts), and -1 is the dropped-token sentinel. Out-of-range IDs remain ignored by the same range checks used for probability reconstruction/kernel routing.

File: tests/test_hybrid_ep.py [hunk 7]
```diff
@@ -168,83 +198,144 @@ def test_hybrid_ep_correctness(buffer: deep_ep.HybridEPBuffer, ref: TorchRef, us
     if dist.get_rank() == 0:
         print('  dispatch+combine API: PASS', flush=True)
 
-    # Dispatch with permute correctness check
-    for fuse_permute_dispatch in [False, True]:
+    # Dense top-k routing dispatch correctness check. Older hybrid-ep branches do
+    # not expose dense topk_idx metadata yet, so skip this block when unsupported.
+    if supports_dense_topk:
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply tests/test_hybrid_ep.py:145]: Addressed in 936e059. The non-permute dispatch+combine correctness coverage now uses the same routing_modes loop as the permute path, covering sparse routing and dense topk_idx with shared dispatch/reference/combine logic.

File: tests/test_hybrid_ep.py [hunk 8]
```diff
@@ -55,6 +56,33 @@ def bitwise_equal(a: torch.Tensor, b: torch.Tensor) -> bool:
...
+    flat_ref = ref.contiguous().view(-1)
+    flat_test = test.contiguous().view(-1)
+    diff_idx = (flat_ref != flat_test).nonzero(as_tuple=True)[0][:5]
+    for idx in diff_idx:
+        i = idx.item()
+        msg += f"\n    [{i}]: ref={flat_ref[i].item()}, got={flat_test[i].item()}"
+    assert False, msg
+
+
+def supports_dense_topk_routing() -> bool:
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply tests/test_hybrid_ep.py:79]: Addressed in 936e059 with a test-side clarification. This helper remains a compatibility gate for tests only; production routing selection now validates dense limits and falls back to sparse routing when needed, so attribute existence is not used as the runtime safety check.

File: csrc/hybrid_ep/config.cuh [hunk 9]
```diff
@@ -113,6 +113,7 @@ struct HybridEpConfigInstance {
   int num_of_ranks_per_node;
   int num_of_nodes;
   int pad_multiple;
+  int topk;  // 0 = sparse bool routing map; >0 = dense int16 topk_idx mode
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply csrc/hybrid_ep/config.cuh]: Addressed in 936e059 with documentation on the config field. TOPK specialization is intentional for dense preprocessing, and the expected K values are bounded by our target use cases; I also re-ran test_hybrid_ep.py on B300 for TOPK=8 after this change.

File: deep_ep/hybrid_ep_buffer.py [hunk 10]
```diff
@@ -186,33 +204,44 @@ def dispatch(
...
-            # Generate the routing map and the probs according to the topk_idx and topk_weights.
-            assert (
-                num_of_experts is not None
-            ), "The number of experts should be provided on index-based routing."
-            if topk_idx is not None:
-                routing_map, probs = indices_to_map(
+            topk = 0
+            routing_data = routing_map
+        elif topk_idx is not None:
+            assert num_of_experts is not None, "num_of_experts is required with topk_idx"
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply deep_ep/hybrid_ep_buffer.py]: Addressed in 936e059. _prepare_routing_data now checks num_of_experts == num_of_experts_per_rank * ranks_per_node * num_nodes for topk_idx calls before constructing probs or launching metadata preprocessing. The same helper is used by dispatch and dispatch_with_permute.

File: deep_ep/hybrid_ep_buffer.py [hunk 11]
```diff
@@ -31,6 +33,22 @@ def indices_to_map(
...
 
+def dense_indices_to_probs(
+    topk_idx: torch.Tensor,
+    topk_weights: torch.Tensor,
+    num_of_tokens: int,
+    num_of_experts: int,
+):
+    probs = torch.zeros(
+        num_of_tokens, num_of_experts, device=topk_idx.device, dtype=torch.float32
+    )
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply deep_ep/hybrid_ep_buffer.py:51]: Addressed in 936e059. Added the distinct-expert assumption comment near dense_indices_to_probs, and updated the sparse topk_idx fallback to use the same valid-mask/scatter_add behavior so malformed duplicate/out-of-range inputs do not diverge silently between fallback and dense reconstruction.

File: csrc/hybrid_ep/executor/executor.cu [hunk 12]
```diff
@@ -29,25 +29,40 @@ torch::Tensor Executor::allgather_routing_map(
...
+    auto num_cols = local_routing_map.size(-1);  // E_total (sparse) or TOPK (dense)
     auto num_of_tokens_per_rank = local_routing_map.size(-2);
     auto group_size = process_group.attr("size")().cast<int>();
-    assert(num_of_expert == config.num_of_experts_per_rank * config.num_of_ranks_per_node * config.num_of_nodes);
+    bool dense_routing = (config.topk > 0);
+    bool custom_allgather_aligned = (local_routing_map.numel() * local_routing_map.element_size()) % 16 == 0;
+    if (!dense_routing) {
+        assert(num_cols == config.num_of_experts_per_rank * config.num_of_ranks_per_node * config.num_of_nodes);
+    } else {
+        assert(num_cols == config.topk);
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply csrc/hybrid_ep/executor/executor.cu:42]: Addressed in 936e059. Added a comment in Executor::allgather_routing_map documenting that custom allgather requires 16B-aligned payloads and dense int16 routing falls back to NCCL when num_tokens * TOPK * sizeof(int16_t) is not aligned.

File: csrc/hybrid_ep/backend/hybrid_ep_backend.cuh [hunk 13]
```diff
@@ -5331,75 +5465,99 @@ __global__ void scan(const bool* input_routing_map,
...
+    // Load routing data for current token (reuse for all ranks).
+    // In dense mode: load TOPK int16 expert indices.
+    // In sparse mode: load E_per_rank * R_per_node bools for local node slice.
+    // Also compute per-rank routing flags.
+    bool token_needed_by_rank_step2[DENSE_ROUTING ? 1 : NUM_OF_RANKS_PER_NODE];
+    // Sparse mode needs the full routing map for local_expert_routing_map writes later.
+    bool token_routing_map[DENSE_ROUTING ? 1 : (NUM_OF_EXPERTS_PER_RANK * NUM_OF_RANKS_PER_NODE)];
+    constexpr int NUM_LOCAL_EXPERT_MASK_WORDS = (NUM_OF_EXPERTS_PER_RANK + 31) / 32;
+    static_assert(!DENSE_ROUTING || NUM_LOCAL_EXPERT_MASK_WORDS <= 16,
+                  "Dense scan supports up to 512 local experts per rank.");
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply csrc/hybrid_ep/backend/hybrid_ep_backend.cuh:5477]: Addressed in 936e059. Python now guards dense routing with the kernel-side limits for local experts per rank and ranks per node. Layouts outside those dense limits fall back to sparse routing instead of reaching the dense kernel static_assert.

**Comment by harryzhou2000** [comment]
Follow-up commit `936e059` addresses the bot review comments around dense top-k routing inference.

Fixes applied:

- Dense routing is no longer selected unconditionally for every `topk_idx` call. `HybridEPBuffer` now uses dense int16 routing only when the dense kernel limits are satisfied: total experts `<= 32768`, local experts per rank `<= 512`, and ranks per node `<= 512`.
- For `topk_idx` layouts outside those dense limits, the Python path now falls back to the previous sparse `indices_to_map` behavior instead of asserting or reaching dense-kernel `static_assert`s. I validated this with a targeted `NUM_LOCAL_EXPERTS=513` distributed check that confirmed the returned handle config uses `topk == 0`.
- `num_of_experts` is now validated against the configured HybridEP layout before constructing probability tensors or launching metadata preprocessing. This prevents mismatches such as a too-narrow `probs` tensor being passed to kernels that index using the configured total expert count.
- The same routing preparation helper is shared by `dispatch()` and `dispatch_with_permute()`, so the validation and dense-vs-sparse fallback behavior are consistent across both APIs.
- The dense expert-id contract is documented at the Python boundary: valid expert IDs are `[0, num_of_experts)`, and `-1` is the dropped-token sentinel. Out-of-range entries are ignored by the existing valid-mask/range-check behavior.
- `indices_to_map()` now also masks invalid expert IDs, so sparse fallback handles `-1` consistently instead of scattering it into the last column.
- Added a short comment documenting the distinct-expert assumption for top-k probabilities. Duplicate expert IDs are malformed input.
- Refactored the non-permute dispatch/combine correctness coverage into the same sparse/dense `routing_modes` loop used by the permute path, removing the duplicated dense-only block.
- Added comments documenting that the test capability check is only a compatibility gate, that dense TOPK is a JIT specialization parameter, and that dense int16 custom allgather falls back to NCCL when the payload is not 16B-aligned.

Validation run on B300 (`umb-b300-025`):

- Rebuilt DeepEP with `PYTORCH_NVCC="ccache nvcc" NVCC_APPEND_FLAGS="--threads 8" TORCH_CUDA_ARCH_LIST="10.0;10.3+PTX" pip install --no-build-isolation . -v`.
- Ran `HIDDEN_DIM=512 NUM_TOKENS_PER_RANK=8192 NUM_LOCAL_EXPERTS=8 TOPK=8 python tests/test_hybrid_ep.py --num-processes 8`.
- BF16 correctness passed, including sparse routing, dense `topk_idx` routing, non-fused permute, fused permute, dense non-fused permute, and dense fused permute.
- FP8 correctness passed for the same coverage.
- Ran a targeted sparse-fallback check with `NUM_LOCAL_EXPERTS=513` and `topk_idx` input; it passed and asserted `handle.config.topk == 0`.
